# Phase 1

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [5]:
import os

# Define the base path for where the shared folder is now located (in My Drive)
my_drive_base_path = '/content/drive/My Drive'

# Construct the full path to the target folder
# Be careful with spaces and special characters, using double quotes around the path components
target_folder_path = os.path.join(
    my_drive_base_path,
    'heta xmer data for ai  (Unzipped Files)'
)

# Check if the folder exists and then change the current working directory
if os.path.exists(target_folder_path):
    os.chdir(target_folder_path)
    print(f"Successfully changed directory to: {os.getcwd()}")
    print("Listing contents of the folder:")
    print(os.listdir())
else:
    print(f"Error: The folder '{target_folder_path}' was not found. Please check the path and sharing settings.")
    print("You might need to adjust 'my_drive_base_path' or the folder names if they are different.")


Successfully changed directory to: /content/drive/.shortcut-targets-by-id/1l2VkwuefAR8cfjJkaLPd-98mpCVWPzqj/heta xmer data for ai  (Unzipped Files)
Listing contents of the folder:
['header.csv', '800 KVA Xmer college', 'JNMC-800 KVA Hostel', 'power_features_analysis.csv', 'power_features_analysis.gsheet', 'header.gsheet', '500 KVA Xmer hospital', 'Notebook']


In [6]:
import pandas as pd
import os

# Load the header from header.csv
header_file_path = 'header.csv'

if os.path.exists(header_file_path):
    header_df = pd.read_csv(header_file_path, header=None)
    header = header_df.iloc[0].tolist()
    print(f"Loaded header with {len(header)} columns.")
else:
    print(f"Error: The file '{header_file_path}' was not found in the current directory: {os.getcwd()}")
    header = [] # Define header as an empty list to prevent NameError in subsequent cells, though processing will still fail

Loaded header with 53 columns.


Next, I'll define a function to read all CSV files within a specified subfolder, apply the loaded header, and combine them into a single DataFrame.

In [7]:
def load_and_process_folder_data(folder_name, header):
    folder_path = os.path.join(os.getcwd(), folder_name)
    all_files = os.listdir(folder_path)
    csv_files = [f for f in all_files if f.endswith('.csv')]

    if not csv_files:
        print(f"No CSV files found in {folder_name}")
        return pd.DataFrame()

    list_dfs = []
    for csv_file in csv_files:
        file_path = os.path.join(folder_path, csv_file)
        try:
            # Read without header, then assign the predefined header
            df = pd.read_csv(file_path, header=None)
            if len(df.columns) == len(header):
                df.columns = header
                list_dfs.append(df)
            else:
                print(f"Skipping {csv_file}: column count mismatch with header ({len(df.columns)} vs {len(header)}).")
        except Exception as e:
            print(f"Error reading {csv_file}: {e}")

    if list_dfs:
        return pd.concat(list_dfs, ignore_index=True)
    else:
        return pd.DataFrame()


Now, let's apply this function to the 'JNMC-800 KVA Hostel', '800 KVA Xmer college', and '500 KVA Xmer hospital' folders.

In [8]:
# Process Hostel data
hostel_df = load_and_process_folder_data('JNMC-800 KVA Hostel', header)
print(f"\nHostel Data - First 5 rows:")
display(hostel_df.head())
print(f"Hostel Data Shape: {hostel_df.shape}")

# Process College data
college_df = load_and_process_folder_data('800 KVA Xmer college', header)
print(f"\nCollege Data - First 5 rows:")
display(college_df.head())
print(f"College Data Shape: {college_df.shape}")

# Process Hospital data
hospital_df = load_and_process_folder_data('500 KVA Xmer hospital', header)
print(f"\nHospital Data - First 5 rows:")
display(hospital_df.head())
print(f"Hospital Data Shape: {hospital_df.shape}")


Hostel Data - First 5 rows:


,Date_time,Host_id,Deviced_id,type,kW_Total,kW_R_phase,kW_Y_phase,kW_B_phase,VAR_Total,VAR_R_phase,...,PF average delivered,Amps average Delivered,Neutral_current,Voltage_R_Harm,Voltage_Y_Harm,Voltage_B_Harm,Current_R_Harm,Current_Y_Harm,Current_B_Harm,kwh deliverd calculated
0,01-01-2025 00:00:38,J4,1,2.0,100327.740,31561.770,36188.527,32577.455,-11679.065,-3443.527,...,0.0.2,0.0.3,29.89933,0.0.4,0.0.5,0.0.6,8.002936,6.963623,7.2806,70531989.162667
1,01-01-2025 00:02:38,J4,1,2.0,94416.760,29082.963,33266.78,32067.008,-11490.223,-3699.3635,...,0.0,0.0,29.27906,0.0,0.0,0.0,8.10632,7.209799,7.235176,70531989.162667
2,01-01-2025 00:04:37,J4,1,2.0,97393.200,28579.207,35346.062,33467.938,-10632.5,-3257.6726,...,0.0,0.0,35.695183,0.0,0.0,0.0,8.662282,6.409847,7.137532,70531989.162667
3,01-01-2025 00:06:37,J4,1,2.0,104120.250,34346.336,35059.344,34714.562,-9454.468,-2789.4087,...,0.0,0.0,23.812328,0.0,0.0,0.0,7.118978,6.376672,6.805122,70531989.162667
4,01-01-2025 00:08:37,J4,1,2.0,102300.266,31742.807,35953.465,34604.0,-10251.382,-2800.523,...,0.0,0.0,31.85061,0.0,0.0,0.0,8.221196,6.528959,6.869356,70531989.162667


Hostel Data Shape: (188639, 53)

College Data - First 5 rows:


,Date_time,Host_id,Deviced_id,type,kW_Total,kW_R_phase,kW_Y_phase,kW_B_phase,VAR_Total,VAR_R_phase,...,PF average delivered,Amps average Delivered,Neutral_current,Voltage_R_Harm,Voltage_Y_Harm,Voltage_B_Harm,Current_R_Harm,Current_Y_Harm,Current_B_Harm,kwh deliverd calculated
0,01-01-2025 00:00:19,AV11,23,2,66997.73,24713.111,18553.506,23731.107,-448.311000,522.2578,...,0.0.2,0.0.3,36.485535,0.0.4,0.0.5,0.0.6,14.524605,18.668898,9.198174,8.526674e+07
1,01-01-2025 00:02:20,AV11,23,2,66823.44,23958.756,19098.314,23766.363,-24.715273,897.2683,...,0.0,0.0,33.751953,0.0,0.0,0.0,14.766725,17.535784,10.069292,8.526674e+07
2,01-01-2025 00:04:20,AV11,23,2,67786.16,24106.379,20597.988,23081.787,1883.718600,1643.0729,...,0.0,0.0,28.90658,0.0,0.0,0.0,13.445116,16.621508,9.938598,8.526674e+07
3,01-01-2025 00:06:20,AV11,23,2,68332.34,23927.670,21165.732,23238.947,2233.055200,1490.4861,...,0.0,0.0,27.980268,1.0410243,0.0,1.5715265,15.093068,16.310678,10.709404,8.526674e+07
4,01-01-2025 00:08:20,AV11,23,2,66865.44,23891.873,19868.973,23104.592,2403.702600,1719.0080,...,0.0,0.0,28.948109,1.5077479,1.0450041,1.1508054,14.664768,16.350120,10.712492,8.526674e+07


College Data Shape: (122962, 53)

Hospital Data - First 5 rows:


,Date_time,Host_id,Deviced_id,type,kW_Total,kW_R_phase,kW_Y_phase,kW_B_phase,VAR_Total,VAR_R_phase,...,PF average delivered,Amps average Delivered,Neutral_current,Voltage_R_Harm,Voltage_Y_Harm,Voltage_B_Harm,Current_R_Harm,Current_Y_Harm,Current_B_Harm,kwh deliverd calculated
0,01-07-2025 00:00:13,AV15,3.0,2.0,130334.0,50014.6,44572.6,35746.9,-4655.81,-1891.03,...,0.0.4,0.0.5,67.5915,1.34782,1.11416,1.08385,3.90963,6.20645,5.56441,3417.294507
1,01-07-2025 00:02:13,AV15,3.0,2.0,126423.0,47085.3,46232.6,33105.1,-7072.56,-3717.94,...,0.0,0.0,61.873,1.21436,1.03757,1.06345,4.1833,4.99761,5.44331,3417.294507
2,01-07-2025 00:04:13,AV15,3.0,2.0,124687.0,47826.5,44594.4,32266.3,-10073.2,-4280.23,...,0.0,0.0,68.9345,1.17271,0.939461,0.947607,4.52372,5.92341,7.13355,3417.294507
3,01-07-2025 00:06:13,AV15,3.0,2.0,136336.0,51332.2,45304.9,39698.9,-5624.11,-3247.63,...,0.0,0.0,56.1291,1.21983,0.918669,1.02479,3.52346,6.47928,5.55556,3417.294507
4,01-07-2025 00:08:14,AV15,3.0,2.0,127018.0,49966.9,40472.6,36578.1,-8697.28,-3051.88,...,0.0,0.0,60.6444,1.28873,1.28695,1.20559,2.95075,6.47668,6.88731,3417.294507


Hospital Data Shape: (170746, 53)


Now, let's aggregate the data for each location on a daily basis. Since each file represents a day, we'll convert the 'Date_time' column to datetime, extract the date, and then sum the total power consumed.

In [9]:
# Function to process and aggregate data by day
def aggregate_by_day(df, name):
    if df.empty:
        print(f"\n{name} DataFrame is empty, skipping daily aggregation.")
        return pd.DataFrame()

    # Convert 'Date_time' to datetime, handling potential errors
    # The format seems to be 'DD-MM-YYYY HH:MM:SS'
    df['Date_time'] = pd.to_datetime(df['Date_time'], format='%d-%m-%Y %H:%M:%S', errors='coerce')

    # Drop rows where Date_time conversion failed
    df.dropna(subset=['Date_time'], inplace=True)

    if df.empty:
        print(f"\n{name} DataFrame became empty after dropping invalid dates, skipping daily aggregation.")
        return pd.DataFrame()

    df['Date'] = df['Date_time'].dt.date

    # Aggregate by Date and sum relevant numerical columns
    # Selecting a few key columns to demonstrate daily aggregation
    daily_agg_df = df.groupby('Date').agg(
        Daily_kW_Total=('kW_Total', 'sum'),
        Daily_kwh_delivered_calculated=('kwh deliverd calculated', 'sum')
    ).reset_index()

    print(f"\n{name} Daily Aggregated Data - First 5 rows:")
    display(daily_agg_df.head())
    print(f"{name} Daily Aggregated Data Shape: {daily_agg_df.shape}")

    return daily_agg_df

# Aggregate Hostel data by day
hostel_daily_df = aggregate_by_day(hostel_df.copy(), 'Hostel')

# Aggregate College data by day
college_daily_df = aggregate_by_day(college_df.copy(), 'College')

# Aggregate Hospital data by day
hospital_daily_df = aggregate_by_day(hospital_df.copy(), 'Hospital')



Hostel Daily Aggregated Data - First 5 rows:


,Date,Daily_kW_Total,Daily_kwh_delivered_calculated
0,2025-01-01,3.232127e+07,50771273877.80883
1,2025-01-02,1.967384e+07,48350948929.955208
2,2025-01-03,2.589668e+07,46493981661.20327
3,2025-01-04,1.729396e+07,47056069253.276207
4,2025-01-05,2.805545e+07,52399241448.860741


Hostel Daily Aggregated Data Shape: (278, 3)

College Daily Aggregated Data - First 5 rows:


,Date,Daily_kW_Total,Daily_kwh_delivered_calculated
0,2025-01-01,4.750604e+06,6.172611e+10
1,2025-01-02,-4.793761e+06,6.260700e+10
2,2025-01-03,-6.431031e+06,6.372057e+10
3,2025-01-04,-1.158078e+07,6.444884e+10
4,2025-01-05,-3.487540e+07,6.576681e+10


College Daily Aggregated Data Shape: (176, 3)

Hospital Daily Aggregated Data - First 5 rows:


,Date,Daily_kW_Total,Daily_kwh_delivered_calculated
0,2025-02-25,4.113967e+07,0.0.290.00.00.00.00.00.00.00.00.00.00.00.00.00...
1,2025-02-26,6.557294e+07,0.0.60.00.00.00.00.00.00.00.00.00.00.00.00.00....
2,2025-02-27,8.563724e+07,18948293.692402
3,2025-02-28,9.899113e+07,19101063.031782
4,2025-03-01,9.872087e+07,19085656.274912


Hospital Daily Aggregated Data Shape: (242, 3)


## Phase 1 — checks (shape, columns, date range)

Per the TODO's "done when" for Phase 1: confirm the schema and a sane date range for each real machine before building anything on top.

In [ ]:
machines_raw = {
    'hostel_J4': hostel_df,
    'college_AV11': college_df,
    'hospital_AV15': hospital_df,
}

for name, df in machines_raw.items():
    print(f"=== {name} ===")
    print("shape:", df.shape)
    print("columns:", df.columns.tolist())
    print("unique Deviced_id:", df['Deviced_id'].unique())
    dt = pd.to_datetime(df['Date_time'], format='%d-%m-%Y %H:%M:%S', errors='coerce')
    print(f"date range: {dt.min()} -> {dt.max()}  ({dt.nunique()} unique timestamps, {dt.isna().sum()} unparsable)")
    print()

In [ ]:
for name, df in machines_raw.items():
    print(f"=== {name}: rows per (Host_id, Deviced_id) ===")
    print(df.groupby(['Host_id', 'Deviced_id']).size().sort_values(ascending=False))
    print()

**Finding:** each location file actually contains a handful of stray rows tagged with the wrong `Deviced_id` (1-4 rows each — almost certainly the same row-corruption issue as the `0.0.2`-style garbage values below, not real extra machines). The real meter per location is the dominant ID: hostel → `Deviced_id 1` (188,622 of 188,639 rows), college → `Deviced_id 23` (122,961 of 122,962), hospital → `Deviced_id 3.0` (170,741 of 170,746). Phase 2 below filters down to that dominant device per location and drops the unparsable-date rows.

## Phase 2 — clean & sanity-check

In [ ]:
# Missing values + duplicate timestamps, on the raw (unfiltered) data
for name, df in machines_raw.items():
    print(f"=== {name}: missing values (nonzero only) ===")
    nulls = df.isnull().sum()
    print(nulls[nulls > 0] if nulls.sum() else "none")
    dup = df.duplicated(subset=['Date_time', 'Deviced_id']).sum()
    print(f"duplicate (Date_time, Deviced_id) rows: {dup}")
    print()

In [ ]:
# Diagnose the garbled '0.0.2'-style values seen in the head() previews, and check
# that the columns we actually need for the pipeline are numerically clean.
suspect_cols = ['PF average delivered', 'Amps average Delivered',
                 'Voltage_R_Harm', 'Voltage_Y_Harm', 'Voltage_B_Harm']
core_cols = ['kW_Total', 'PF_Ave', 'Current_Total', 'Current_R_phase', 'Current_Y_phase',
             'Current_B_phase', 'Current_R_Harm', 'Current_Y_Harm', 'Current_B_Harm', 'Neutral_current']

for name, df in machines_raw.items():
    print(f"=== {name} ===")
    for col in suspect_cols:
        bad = pd.to_numeric(df[col], errors='coerce').isna().sum()
        print(f"  [unused col] {col}: {bad}/{len(df)} non-numeric ({bad/len(df):.1%})")
    for col in core_cols:
        bad = pd.to_numeric(df[col], errors='coerce').isna().sum()
        if bad:
            print(f"  [PIPELINE col] {col}: {bad}/{len(df)} non-numeric")
    neg_kw = (pd.to_numeric(df['kW_Total'], errors='coerce') < 0).sum()
    print(f"  kW_Total negative readings: {neg_kw}/{len(df)} ({neg_kw/len(df):.1%})")
    print()

In [ ]:
import matplotlib.pyplot as plt

dominant_device = {'hostel_J4': 1, 'college_AV11': 23, 'hospital_AV15': 3.0}

fig, axes = plt.subplots(3, 3, figsize=(18, 10))
for row, (name, df) in enumerate(machines_raw.items()):
    d = df[df['Deviced_id'] == dominant_device[name]].copy()
    d['Date_time'] = pd.to_datetime(d['Date_time'], format='%d-%m-%Y %H:%M:%S', errors='coerce')
    d = d.dropna(subset=['Date_time']).sort_values('Date_time')
    for col, ax in zip(['kW_Total', 'PF_Ave', 'Current_R_Harm'], axes[row]):
        ax.plot(d['Date_time'], pd.to_numeric(d[col], errors='coerce'), lw=0.3)
        ax.set_title(f"{name} — {col}")
        ax.tick_params(axis='x', rotation=30)
plt.tight_layout()
plt.savefig('/tmp/phase2_raw_signals.png', dpi=100)
plt.show()

**Load indicator decision:** `Current_Total` (per the TODO's recommendation) — it's non-negative and avoids the `kW_Total` sign anomaly flagged above.

## Phase 3 — build features (preprocessing)

Real readings come in every ~2 minutes, not daily, so before the 7-day rolling slope from the prototype notebook (§3) makes sense we resample each machine to one row per day. Steps: filter to the dominant device per location, coerce pipeline columns to numeric, resample to daily means, compute phase imbalance %, normalize the harmonics/neutral current by the load indicator, then run the prototype's exact `rolling_slope` / `add_trend_features` against it.

In [ ]:
pipeline_cols = ['kW_Total', 'PF_Ave', 'Current_Total', 'Current_R_phase', 'Current_Y_phase',
                  'Current_B_phase', 'Current_R_Harm', 'Current_Y_Harm', 'Current_B_Harm', 'Neutral_current']

def prep_machine_daily(df, device_id, machine_id):
    d = df[df['Deviced_id'] == device_id].copy()
    d['Date_time'] = pd.to_datetime(d['Date_time'], format='%d-%m-%Y %H:%M:%S', errors='coerce')
    for col in pipeline_cols:
        d[col] = pd.to_numeric(d[col], errors='coerce')
    d = d.dropna(subset=['Date_time']).sort_values('Date_time')
    daily = d.set_index('Date_time')[pipeline_cols].resample('D').mean()
    daily['rows_that_day'] = d.set_index('Date_time')['kW_Total'].resample('D').count()
    daily['machine_id'] = machine_id
    return daily.reset_index().rename(columns={'Date_time': 'date'})

daily_hostel = prep_machine_daily(hostel_df, 1, 'hostel_J4')
daily_college = prep_machine_daily(college_df, 23, 'college_AV11')
daily_hospital = prep_machine_daily(hospital_df, 3.0, 'hospital_AV15')

daily_all = pd.concat([daily_hostel, daily_college, daily_hospital]).reset_index(drop=True)
print(daily_all.shape)
print(daily_all.groupby('machine_id')['date'].agg(['min', 'max', 'count']))
daily_all.head()

In [ ]:
# Phase imbalance % and normalization by the load indicator (Current_Total)
phases = daily_all[['Current_R_phase', 'Current_Y_phase', 'Current_B_phase']]
daily_all['phase_imbalance_pct'] = (
    (phases.max(axis=1) - phases.min(axis=1)) / phases.mean(axis=1) * 100
)

for col in ['Current_R_Harm', 'Current_Y_Harm', 'Current_B_Harm', 'Neutral_current']:
    daily_all[f'{col}_norm'] = daily_all[col] / daily_all['Current_Total']

daily_all[['machine_id', 'date', 'Current_Total', 'phase_imbalance_pct',
           'Current_R_Harm_norm', 'Neutral_current_norm']].head()

In [ ]:
# From the prototype notebook §3 (Predictive_Maintenance_Prototype.ipynb), unchanged logic —
# run against the real, resampled daily_all instead of synthetic data.
import numpy as np

def rolling_slope(series, window):
    slopes = []
    for i in range(len(series)):
        if i < window - 1:
            slopes.append(np.nan)
            continue
        y = series.iloc[i - window + 1:i + 1].values
        x = np.arange(window)
        slopes.append(np.polyfit(x, y, 1)[0])
    return pd.Series(slopes, index=series.index)

def add_trend_features(g, window=7):
    g = g.sort_values('date').copy()
    for col in ['Current_R_Harm_norm', 'Current_Y_Harm_norm', 'Current_B_Harm_norm',
                'PF_Ave', 'Neutral_current_norm', 'phase_imbalance_pct']:
        g[f'{col}_slope_{window}d'] = rolling_slope(g[col], window)
    return g

# loop per machine explicitly (see prototype note: groupby().apply() can silently drop
# the group key column in some pandas versions when the function returns the full frame)
daily_feat = pd.concat([add_trend_features(g) for _, g in daily_all.groupby('machine_id')]).reset_index(drop=True)

slope_cols = [c for c in daily_feat.columns if c.endswith('_slope_7d')]
print("non-null slope rows per machine:")
print(daily_feat.groupby('machine_id')[slope_cols[0]].apply(lambda s: s.notna().sum()))
daily_feat[['machine_id', 'date'] + slope_cols].dropna().head()

Spot-check: slope values are non-zero and vary day to day rather than sitting flat at zero or NaN — matches the "done when" bar for Phase 3.

## Phase 4 — Health Index (Stage A)

**Baseline note:** no manufacturer spec was available, so `BASELINE` below is estimated empirically from this fleet's own full-year history (mean/std per feature, across all three machines, since these are load-normalized ratios that should be comparable across differently-sized transformers). **Flag as "to confirm" — replace with manufacturer/OEM spec numbers if you get them**, per the TODO.

In [ ]:
# Population-prior baseline — TO CONFIRM. Estimated from this fleet's own history since no
# manufacturer spec was supplied (see prototype notebook §4 for the production-intent version,
# where this comes from Layer 0 config / Unit 3's Stage 1 priors instead).
baseline_source_cols = ['Current_R_Harm_norm', 'Current_Y_Harm_norm', 'Current_B_Harm_norm',
                         'PF_Ave', 'Neutral_current_norm', 'phase_imbalance_pct']
BASELINE = {col: (daily_all[col].mean(), daily_all[col].std()) for col in baseline_source_cols}
for col, (mean, std) in BASELINE.items():
    print(f"  {col}: mean={mean:.4f}, std={std:.4f}")

def compute_health_index(row):
    z_scores = [abs(row[col] - mean) / std for col, (mean, std) in BASELINE.items()]
    avg_z = np.mean(z_scores)
    return max(0, 100 - avg_z * 12)

daily_feat['health_index'] = daily_feat.apply(compute_health_index, axis=1)
daily_feat.groupby('machine_id')['health_index'].agg(['min', 'max', 'mean'])

In [ ]:
fig, ax = plt.subplots(figsize=(12, 5))
for mid, g in daily_feat.groupby('machine_id'):
    g = g.sort_values('date')
    ax.plot(g['date'], g['health_index'], label=mid, lw=1)
ax.axhline(50, color='red', linestyle='--', label='Example risk threshold')
ax.set_ylabel('Health index (0-100)')
ax.set_title('Health index over time — real fleet data')
ax.legend()
plt.tight_layout()
plt.savefig('/tmp/health_index_chart.png', dpi=150)
plt.show()

Pipeline runs cleanly on real data and produces a Health Index (0–100) for every machine every day it has data — matches Phase 4's "done when." `college_AV11` sits persistently lower (~48 average) than the other two (~82–84); given no machine is known to be failing yet, this most likely reflects `college_AV11`'s own baseline deviation (its fleet-wide `PF_Ave` z-score is larger) rather than genuine degradation — worth a second look once you have real maintenance history to check against, per the note below.

## Phase 5 — early-warning detection (CUSUM)

In [ ]:
# From the prototype notebook §5, unchanged, run against each real machine's health_index series.
def cusum_flags(series, threshold=8, drift=0.5):
    s_pos, s_neg, flags = 0, 0, []
    baseline_mean = series.dropna().iloc[:10].mean()
    for val in series.fillna(method='ffill'):
        s_pos = max(0, s_pos + (baseline_mean - val) - drift)
        s_neg = min(0, s_neg + (baseline_mean - val) + drift)
        flags.append(s_pos > threshold or abs(s_neg) > threshold)
    return flags

for mid, g in daily_feat.groupby('machine_id'):
    g = g.sort_values('date')
    flags = cusum_flags(g['health_index'])
    n_flags = sum(flags)
    first_flag = next((i for i, f in enumerate(flags) if f), None)
    when = g['date'].iloc[first_flag] if first_flag is not None else 'never'
    print(f"{mid}: first change-point flagged at {when}  ({n_flags}/{len(flags)} days flagged total)")

In [ ]:
# Why is CUSUM over-flagging? Real health_index is noisier day-to-day than the prototype's
# smooth synthetic 60-day series, so default threshold=8/drift=0.5 (tuned for that smooth
# signal) trips almost immediately on real data.
for mid, g in daily_feat.groupby('machine_id'):
    g = g.sort_values('date')
    hi = g['health_index']
    print(f"{mid}: health_index day-to-day std={hi.diff().std():.2f}, "
          f"overall std={hi.std():.2f}, first-10-day baseline mean={hi.iloc[:10].mean():.2f}")

**Finding:** the flags are explainable, but not clean — day-to-day health-index std (8–26) is on the same order as `threshold=8`, so CUSUM (tuned in the prototype for a smooth 60-day synthetic series) trips almost every day on real, noisier data. Function runs without error, satisfying the letter of Phase 5's "done when," but the defaults need re-tuning (larger `threshold`/`drift`, or smoothing `health_index` first) before this is a useful early-warning signal rather than a constant alarm. **Flag as "to confirm"** alongside the Phase 4 baseline.

## Phase 6 — package the output & prep to demo

In [ ]:
# From the prototype notebook §7, adapted only for the real (normalized) feature column names.
import uuid

def predict_maintenance_risk(machine_id, latest_row, confidence_stage='stage_a_unsupervised'):
    health_index = latest_row['health_index']
    risk_tier = 'high' if health_index < 40 else 'medium' if health_index < 65 else 'low'
    return {
        "event_id": str(uuid.uuid4()),
        "machine_id": machine_id,
        "timestamp": str(latest_row['date']),
        "source_unit": "predictive_maintenance",
        "event_type": "maintenance_risk",
        "severity": "alert" if risk_tier == "high" else "info",
        "health_index": round(float(health_index), 1),
        "risk_tier": risk_tier,
        "confidence_stage": confidence_stage,
        "contributing_features": {
            "current_r_harm_norm_slope_7d": round(float(latest_row['Current_R_Harm_norm_slope_7d']), 4),
            "pf_ave_slope_7d": round(float(latest_row['PF_Ave_slope_7d']), 4),
            "phase_imbalance_slope_7d": round(float(latest_row['phase_imbalance_pct_slope_7d']), 3),
        },
    }

import json
for mid, g in daily_feat.groupby('machine_id'):
    latest = g.dropna(subset=['health_index', 'Current_R_Harm_norm_slope_7d']).sort_values('date').iloc[-1]
    result = predict_maintenance_risk(mid, latest)
    print(json.dumps(result, indent=2))
    print()

Health Index chart was saved to `/tmp/health_index_chart.png` in Phase 4 — that's the core visual for a demo.

**Why each machine's risk tier is what it is (latest reading, 2025-11-11):**
- **hostel_J4 — low (91.0):** all six trend slopes are near zero — harmonics, PF, and phase imbalance are essentially flat, i.e. a stable machine.
- **hospital_AV15 — low (86.4):** currents, PF, and phase balance all sit close to fleet-normal, with only a mild PF/imbalance drift.
- **college_AV11 — low (68.4)**, but closest to the medium boundary: its PF_Ave has been trending *up* recently (slope +0.11), pulling the index up from a much noisier, historically lower baseline (~48 average) — worth watching rather than treating as clearly healthy.

This closes out the checklist in `Predictive_Maintenance_Colab_TODO.md`: a real JSON payload from real data, a presentable chart, and a one-line reason per machine. Two things are explicitly flagged **"to confirm"** before trusting this beyond a pipeline demo: the Phase 4 `BASELINE` (currently self-derived from this fleet, not a manufacturer spec) and the Phase 5 CUSUM thresholds (currently over-flagging on real data's day-to-day noise).

# Hospital

In [ ]:
print('hospital_df' in dir(), 'daily_feat' in dir())
h = hospital_df[hospital_df['Deviced_id'] == 3.0].copy()
print(h.shape)

In [ ]:
device_channels = {
    'Active power (kW) meter': ['kW_Total', 'kW_R_phase', 'kW_Y_phase', 'kW_B_phase'],
    'Reactive power (VAR) meter': ['VAR_Total', 'VAR_R_phase', 'VAR_Y_phase', 'VAR_B_phase'],
    'Apparent power (VA) meter': ['VA_total', 'VA_R_phase', 'VA_Y_phase', 'VA_B_phase'],
    'Power factor (PF) meter': ['PF_Ave', 'PF_R_phase', 'PF_Y_phase', 'PF_B_phase',
                                 'PF average Received', 'PF average delivered'],
    'Voltage sensor (PT)': ['VLL_average', 'Vry_phase', 'Vyb_phase', 'Vbr_phase',
                             'VLN_average', 'V_R_phase', 'V_Y_phase', 'V_B_phase'],
    'Current sensor (CT)': ['Current_Total', 'Current_R_phase', 'Current_Y_phase', 'Current_B_phase',
                             'Amps average Received', 'Amps average Delivered', 'Neutral_current'],
    'Frequency meter': ['Frequency'],
    'Energy meter (cumulative registers)': ['kWh_Received', 'VAh_Received', 'VARh_Ind._Received',
                                             'VARh_Cap._Received', 'kWh_Delivered', 'VAh_Delivered',
                                             'VARh_Ind._Delivered', 'VARh_Cap._Delivered',
                                             'kwh deliverd calculated'],
    'Harmonic analyzer': ['Voltage_R_Harm', 'Voltage_Y_Harm', 'Voltage_B_Harm',
                           'Current_R_Harm', 'Current_Y_Harm', 'Current_B_Harm'],
}
metadata_cols = ['Date_time', 'Host_id', 'Deviced_id', 'type']

accounted = sum(device_channels.values(), []) + metadata_cols
print("columns accounted for:", len(accounted), "/ 53")
print("missing from grouping:", set(header) - set(accounted))

for group, cols in device_channels.items():
    print(f"\n=== {group} ===")
    for col in cols:
        s = pd.to_numeric(h[col], errors='coerce')
        print(f"  {col}: min={s.min():.3g}, mean={s.mean():.3g}, max={s.max():.3g}")

## Devices / sensor channels we have

The hospital's real device (`Deviced_id 3.0`, the 500 KVA transformer meter — 170,741 of 170,746 rows) is a **single multi-function power-quality meter**, not multiple physical devices. What looks like "many devices" in `header.csv` is really this one meter's internal measurement channels, each fed by its own sensor (CT/PT) or computed register. Grouped by channel, with real observed ranges from the hospital's own data pulled above:

| Channel (sensor) | Columns | Functionality | Why it matters for a hospital transformer | What we can predict / analyze |
|---|---|---|---|---|
| **Active power (kW) meter** | `kW_Total`, `kW_R/Y/B_phase` | Real power actually doing work, per phase and total. Observed: Total 0–356 kW, phases individually swing negative (–39 kW) even though Total floors at 0. | Hospitals run life-support, HVAC, sterilization, imaging — sustained overload directly threatens uptime. Per-phase negatives hint at reverse flow or sign-convention noise on individual phases even when net flow is import. | Overload risk, load-growth trend (rising baseline kW over months → capacity planning), whether the transformer is being pushed toward its 500 KVA nameplate limit. |
| **Reactive power (VAR) meter** | `VAR_Total`, `VAR_R/Y/B_phase` | Non-working power exchanged with the grid (inductive motors, MRI/imaging equipment, etc.). | High/volatile VAR draw strains the transformer and utility supply without doing useful work; often the first thing to shift before a PF-correction capacitor bank fails. | Health of any PF-correction equipment upstream; whether reactive loading is trending in a way that erodes effective capacity. |
| **Apparent power (VA) meter** | `VA_total`, `VA_R/Y/B_phase` | Vector sum of kW + VAR — what the transformer is actually sized against (its 500 KVA rating is a VA rating). | Direct comparison to nameplate capacity; VA_total ÷ 500,000 is the real "how loaded is this transformer" number. | Capacity headroom, whether load has room to grow before the transformer needs upsizing. |
| **Power factor (PF) meter** | `PF_Ave`, `PF_R/Y/B_phase`, `PF average Received/delivered` | Ratio of real to apparent power (0–1). Observed: `PF_Ave` mean **0.503** — a genuinely poor average PF for this site. | Poor/degrading PF means the transformer works harder to deliver the same useful power, raising internal heating and wear for a given load. A hospital running PF ~0.5 is leaving real thermal margin on the table. | Whether PF is degrading over time (a slow-signal precursor the TODO's Phase 3 already tracks via `PF_Ave_slope_7d`); a concrete, actionable finding (PF correction) independent of fault prediction. |
| **Voltage sensor (PT)** | `VLL_average`, `Vry/yb/br_phase`, `VLN_average`, `V_R/Y/B_phase` | Line-line and line-neutral voltage per phase. Observed: VLL ~420 V, VLN ~242 V — consistent with a 415/240 V distribution transformer, sane. | Sustained over/under-voltage stresses winding insulation; voltage sags during high-demand periods (e.g. imaging equipment startup) can indicate the transformer is undersized or tap position is wrong. | Insulation stress trending, tap-changer misconfiguration, whether voltage dips correlate with specific hospital equipment cycles. |
| **Current sensor (CT)** | `Current_Total`, `Current_R/Y/B_phase`, `Amps average Received/Delivered`, `Neutral_current` | Load current per phase plus neutral return current. This is the TODO's chosen **load indicator** (`Current_Total`). | Neutral current should be small on a balanced 3-phase load; a rising neutral current is a classic early sign of load imbalance or a developing winding/connection fault. | Phase imbalance trending (already in the pipeline), neutral-current drift as an independent early-warning signal, load profile shape (day/night, weekday/weekend cycles typical of a hospital). |
| **Frequency meter** | `Frequency` | Grid frequency at the meter. Observed: mostly ~50 Hz as expected, but with 0 and 63.5 Hz outliers (see data-quality note below). | Frequency excursions usually indicate grid instability or a meter comms dropout, not the transformer itself — but sudden dropouts during a reading is useful as a **data-availability** signal for a critical hospital feed. | Comms/sensor uptime monitoring — for a hospital, knowing when the meter itself went dark is as important as knowing what it measured. |
| **Energy meter (cumulative registers)** | `kWh_Received`, `VAh_Received`, `VARh_Ind./Cap._Received`, `kWh/VAh/VARh_Delivered`, `kwh deliverd calculated` | Cumulative lifetime energy counters. Observed: the `..._Delivered` registers are **always exactly 0** — dead/unwired on this meter; `kwh deliverd calculated` (a derived field) is the only live "delivered energy" signal. | Not directly a health signal, but billing/consumption verification and a sanity cross-check against the kW readings (energy delta over time should track kW × time). | Consumption trend for capacity planning; NOT currently useful for fault detection. |
| **Harmonic analyzer** | `Voltage_R/Y/B_Harm`, `Current_R/Y/B_Harm` | Total harmonic distortion on voltage and current. Observed: voltage THD ~1% (clean), current THD averages 5–7% but spikes to 500–700% (implausible — see data-quality note). | Hospitals run a lot of switch-mode/nonlinear loads (imaging, UPS, variable-speed drives) that inject harmonics; sustained high current THD causes extra transformer heating (skin/proximity effect) beyond what the RMS current alone would suggest. | Already the pipeline's primary early-warning feature (`Current_R/Y/B_Harm_slope_7d`) — this channel is the closest thing to a leading indicator we have today. |

## Other parameters we have (not sensor readings)

Apart from the ten channels above, `header.csv` also carries **metadata/identity fields** that are not physical measurements themselves but are required to use the data at all:

| Column | What it is | Role |
|---|---|---|
| `Date_time` | Reading timestamp | Join key / time axis for every trend feature in Phases 3–5 |
| `Host_id` | Site code (`AV15` for this hospital) | Distinguishes location when combining fleets |
| `Deviced_id` | Meter/device ID within a site | Distinguishes the real meter (`3.0`) from the stray corrupted rows filtered out in Phase 1/2 |
| `type` | A numeric meter-type code (constant `2.0` for this device) | Currently unused — worth asking the vendor what the code space means; could matter once more meter types are onboarded |

## Data-quality flags found in this pass (worth fixing before Health Index trusts these channels)

Pulling real min/mean/max per channel surfaced issues beyond the ones already in Phase 2:
- **Dead registers**: `PF average delivered`, `Amps average Delivered`, `kWh_Delivered`, `VAh_Delivered`, `VARh_Ind./Cap._Delivered` are constant **0** for the entire year — not wired on this meter. `PF average Received` is constant ~-0.945 — effectively frozen, not a live signal either. None of these should be fed into the Health Index or trend features; the current pipeline already avoids them.
- **Frequency dropouts**: `Frequency` ranges from **0 to 63.5 Hz**, both physically implausible for a 50 Hz grid — these are almost certainly comms/sensor dropouts, not real events. Should be clipped to a sane band (e.g. 47–53 Hz) and treated as missing outside it.
- **Harmonic outliers**: `Current_R/Y/B_Harm` average 5–7% (plausible) but spike as high as **500–710%** — not physically real THD values. These extreme spikes should be capped or filtered (e.g. drop values above ~100%) before they're allowed to swing the Health Index or the rolling slope in Phase 3.

## What's still needed for a more complete Layer 3 health analysis

Everything above comes from one electrical meter. That's enough to prove the Stage A pipeline (Phases 1–6), but real transformer condition monitoring leans heavily on signals this meter simply doesn't have. Roughly in order of predictive value for a transformer specifically (vs. the load it serves):

**Highest-value gap — thermal/insulation condition:**
- **Top-oil and winding-hotspot temperature (OTI/WTI)** — the single most direct transformer health signal there is; almost every industry-standard predictive model for transformers is built around temperature rise vs. load, not electrical signature alone.
- **Dissolved Gas Analysis (DGA)**, even periodic lab samples rather than continuous online sensors — this is the gold-standard signal for detecting incipient insulation breakdown, arcing, and overheating *before* it becomes an electrical symptom. Nothing in the current dataset substitutes for this.
- **Oil level and oil dielectric strength / moisture (BDV/PPM tests)** — moisture ingress is a leading cause of insulation failure and isn't visible in any electrical channel here.

**Mechanical / auxiliary systems:**
- **Tap-changer (OLTC) position and operation count** — a high-wear mechanical component that fails independently of the windings; not observable electrically until it's already causing voltage problems.
- **Cooling system status** (fan/pump run state, cooling mode ONAN/ONAF) — without this, a temperature rise can't be distinguished from "normal, cooling just kicked in" vs. "cooling has failed."
- **Bushing monitoring** (capacitance/tan-delta, partial discharge) — bushings are a common hospital-transformer failure point and are invisible to the panel meter.
- **Breather/silica-gel condition** — cheap, simple, and a real leading indicator of moisture ingress that's currently not logged anywhere.
- **Vibration/acoustic sensors** — useful for core looseness or mechanical winding deformation, again invisible electrically until advanced.

**Context data (not new sensors, but needed to interpret what we already have):**
- **Nameplate rating and load history in % of rated capacity**, not just raw amps — `Current_Total` alone doesn't say how close to the transformer's actual limit the machine is running.
- **Ambient temperature / weather** — needed to separate "hot because it's summer" from "hot because it's degrading." Could be pulled from an external weather API keyed on `Date_time` rather than a new physical sensor.
- **Protection-relay trip/alarm event log** (Buchholz relay, differential/overcurrent trips) — this is exactly the real-world label source Stage B needs (per `Predictive_Maintenance_Implementation_Plan.md`), and it's currently the single biggest gap for turning this from an unsupervised pipeline into a validated one.
- **Maintenance/inspection log** — same gap as above; without at least 10-20 real logged events, there's no way to check whether the Health Index or CUSUM flags actually mean anything for *this* fleet.
- **Criticality weighting for the risk tiers** — this is a hospital: an identical Health Index reading here probably warrants a higher-urgency response than the same reading at the college. Worth a site-criticality multiplier once Layer 4/5 consume these `maintenance_risk` events, so alerting doesn't treat all three sites as equally urgent.